### DISCUSSION TITLE
How to create an environment development for working into package flutter

### DISCUSSION THREAD

**Discussion by darkomenx** [doonfrs/pluto_grid_plus#1] [General]
Hello everybody,

First, thank you for your fork and your work to bugfix and maintain and upgrade this good package datatable.

In my part, I also need to add an feature on this PlutoGrid that not exist or don't find possibility to do my requirement.

**I would this :** 

<img width="1108" alt="example" src="https://github.com/doonfrs/pluto_grid_plus/assets/16452686/0c1af255-119f-4e65-991a-adad8a48d4ee">

**resume :** add possibility to add items into a popup datatable items

But I don't know how developing just a package flutter. Because with an application I have screen to testing my change and only after I commit my change but into package project.....

- I don't have screen for testing ?
- Does I need to create functional testing with simulated screen with datatable ?
- what is the structure of an Android studio project for working into package project ?

---------------------------------------------------------------------------------------------------------------

In concrete example code 

into this PlutoGrid declaration column code, we can't use another things that String type to display information into popup menu cell

```
PlutoColumn(
        title: 'Category',
        field: 'category',
        readOnly: false,
        type: PlutoColumnType.select(listCategoryLabel, defaultValue: '', enableColumnFilter: true),
    ),
```

Perhaps I think that I can change this function for resolve the my requirement into the pluto_grid_popup.dart file 

```
`Future<void> open() async {
    final textDirection = Directionality.of(context);

    final borderRadius = configuration.style.gridBorderRadius.resolve(
      textDirection,
    );

    PlutoGridOnSelectedEvent? selected =
        await showDialog<PlutoGridOnSelectedEvent>(
            context: context,
            builder: (BuildContext ctx) {
              return Dialog(
                shape: borderRadius == BorderRadius.zero
                    ? null
                    : RoundedRectangleBorder(borderRadius: borderRadius),
                child: LayoutBuilder(
                  builder: (ctx, size) {
                    return SizedBox(
                      width: (width ?? size.maxWidth) +
                          PlutoGridSettings.gridInnerSpacing,
                      height: height ?? size.maxHeight,
                      child: Directionality(
                        textDirection: textDirection,
                        child: PlutoGrid(
                          columns: columns,
                          rows: rows,
                          columnGroups: columnGroups,
                          onLoaded: onLoaded,
                          onChanged: onChanged,
                          onSelected: (PlutoGridOnSelectedEvent event) {
                            Navigator.pop(ctx, event);
                          },
                          onSorted: onSorted,
                          onRowChecked: onRowChecked,
                          onRowDoubleTap: onRowDoubleTap,
                          onRowSecondaryTap: onRowSecondaryTap,
                          onRowsMoved: onRowsMoved,
                          onColumnsMoved: onColumnsMoved,
                          createHeader: createHeader,
                          createFooter: createFooter,
                          noRowsWidget: noRowsWidget,
                          rowColorCallback: rowColorCallback,
                          columnMenuDelegate: columnMenuDelegate,
                          configuration: configuration,
                          mode: mode,
                        ),
                      ),
                    );
                  },
                ),
              );
            });
    if (onSelected != null && selected != null) {
      onSelected!(selected);
    }
  }`
```

I hope that understand my missunderstanding and thank you for your advice.

**Comment by doonfrs** [comment]
I hope I understood your request well, I've added a new feature allowing the developer to pass a builder function to control the output widget for the select instead of regular text

New Version: 8.2.0

Notice the builder function ( optional ) 
```
PlutoColumn(
      title: 'Role 2',
      field: 'role2',
      type: PlutoColumnType.select(
        <String>[
          'Programmer',
          'Designer',
          'Owner',
        ],
        builder: (item) {
          return Row(children: [
            Icon(item == 'Programmer' ? Icons.code : Icons.design_services),
            const SizedBox(width: 8),
            Text(item),
          ]);
        },
      ),
    ),
```

![image](https://github.com/doonfrs/pluto_grid_plus/assets/4661009/0ee052a2-2665-4607-8422-d211ece2d7e8)

**Reply by darkomenx** [reply]
Oh my god what's fastly features develop and response :) 

In my context, this option is good but not filling my requirement because each row bring it's own options. I need option that dealing with object type (not a row/instance) because it's for create new instance of this type.
If I use your feature for added Icon.onPressed() function then each row have possibilty to add new item. Perhaps that run correctly but in context of UI/UX it's strange.

The best it's just have one latest line/row with this widget (for added items).

Your feature not corresponding in my requirement BUT it can be fitted very well for lot of another things.
I think that you need to keep your work !

Do you think that is easy to added last row with widget actions ?
Or tell me how work on your fork if you want.

I hope that more understand me AND very apologize it's my demonstration was not very clear.

Have a good day !
